### DISCUSSION TITLE
Possible to generate class from json schema definition?

### DISCUSSION THREAD

**Discussion by CaptainDriftwood** [wyfo/apischema#206] [Q&A]
Given a json schema definition, is it possible to generate a python dataclass (model) from that definition?

**Comment by wyfo** [comment]
Hi, there is no code generation feature in *apischema*, and it's not really planned at the moment.

I've already thought about it, but I've put it aside for several reasons:
- Models are often ORM classes (because they can be easily used as-is in *apischema*), which makes code generation a lot more complex; it depends of the ORM (SQLAlchemy, MongoEngine, etc.), and a lot of database specific things are not necessarily expressed in JSON schema (indexes, relation vs. array in Postgresql, etc.). However, last version of SQLAlchemy can now use dataclasses, so things may evolve. 
- There can be more than one way to generate code, for example, [composition vs. inheritance](https://wyfo.github.io/apischema/data_model/#composition-over-inheritance-composed-dataclasses-flattening), conversion registration (for example `bson.ObjectId`) should be handled too, etc.
- Personally, I don't really like this kind of code generation (and I had a bad professional experience with it); it can be nice to have to start a project, but when it comes to updates, it's an other thing. Regenerating models can break the code, while renaming a field is easy when using IDE refactoring tools. I know that having a common schema can help synchronization between services (and product design), but I rather prefer to generate the schema from the code and validate it (in an automated way in the CI) according to the common schema.

Also, code generation subject seems to be quite independent of *apischema*, as I don't see which part of the code I would reuse for this feature. So I think it should rather be implemented in an other project.

**Reply by CaptainDriftwood** [reply]
I definitely think there would be value in having this feature, even if it did not reside in apischema itself, but rather an additional library to use if a user wanted to. Fingers crossed we could get something fleshed out for this down the road.
